File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = set()
        nums.sort()
        for i in range (len(nums)-2):
            if i>0 and nums[i]==nums[i]-1: 
                continue

            left , right = i+1, len(nums)-1
            while left<right:
                sumtotal = nums[i] + nums[left] + nums[right]

                if sumtotal<0: left = left+1
                elif sumtotal>0: right = right -1 
                else: 
                    res.add((nums[i],nums[left],nums[right]))
                    while left<right and nums[left]==nums[left+1]: left = left+1
                    while left<right and nums[right]==nums[right-1]: right = right-1

                    left = left+1 
                    right = right-1
        return res

File: 0015-3sum/0015-3sum.cpp
class Solution {
public:
    vector<vector<int>> result;

    void twoSum(vector<int>& nums, int target, int i, int j) {
        while (i < j) {
            int sum = nums[i] + nums[j];
            if (sum > target) {
                j--;
            } else if (sum < target) {
                i++;
            } else {
                result.push_back({-target, nums[i], nums[j]});
                while (i < j && nums[i] == nums[i + 1]) i++;
                while (i < j && nums[j] == nums[j - 1]) j--;
                i++;
                j--;
            }
        }
    }

    vector<vector<int>> threeSum(vector<int>& nums) {
        int n = nums.size();
        result.clear();
        if (n < 3) return {};

        sort(nums.begin(), nums.end());

        for (int i = 0; i < n - 2; i++) {
            if (i > 0 && nums[i] == nums[i - 1]) continue;  // skip duplicate n1
            int n1 = nums[i];
            int target = -n1;
            twoSum(nums, target, i + 1, n - 1);
        }
        return result;
    }
};
